File: 399. Evaluate Division/solution.py
class Solution(object):
    def __init__(self):
        self.graph = {}
        
    def calcEquation(self, equations, values, queries):
        """
        :type equations: List[List[str]]
        :type values: List[float]
        :type queries: List[List[str]]
        :rtype: List[float]
        """
        # create a graph representation
        for x, y in equations:
            self.graph[x] = {x:1.0}
            self.graph[y] = {y:1.0}
            
        for i in range(len(equations)):
            x = equations[i][0]
            y = equations[i][1]
            self.graph[x][y] = values[i]
            self.graph[y][x] = 1/values[i]
            
        # perform dfs for each query
        sol = []
        for x, y in queries:
            if not x in self.graph or not y in self.graph:
                sol.append(-1.0)
            else:
                visited = set()
                tmp = self.dfs(visited, x, y)
                if tmp == None:
                    sol.append(-1.0)
                else:
                    sol.append(tmp)
        return sol
    
    def dfs(self, visited, x, y):
        visited.add(x)
        if y in self.graph[x]:
            return self.graph[x][y]
        
        for z in self.graph[x]:
            if not z in visited:
                sol = self.dfs(visited, z, y)
                if sol != None:
                    self.graph[x][y] = sol*self.graph[x][z]
                    return self.graph[x][y]
                
        return None

File: 399. Evaluate Division/solution.java
class Solution {
    public double[] calcEquation(List<List<String>> equations, double[] values, List<List<String>> queries) {
        // create graph
        Map<String, Map<String, Double>> graph = new HashMap<String, Map<String, Double>>();
        HashSet<String> vertices = new HashSet<String>();
        int n = values.length;
        List<String> equation;
        Map<String, Double> adjList;
        
        for (int i = 0; i < n; i++) {
            equation = equations.get(i);
            
            // update vertices
            vertices.add(equation.get(0));
            vertices.add(equation.get(1));
            
            // insert self
            if (!graph.containsKey(equation.get(0))) {
                adjList = new HashMap<String, Double>();
                adjList.put(equation.get(0), 1.0);
                graph.put(equation.get(0), adjList);
            }
            
            if (!graph.containsKey(equation.get(1))) {
                adjList = new HashMap<String, Double>();
                adjList.put(equation.get(1), 1.0);
                graph.put(equation.get(1), adjList);
            }
        }
        
        // update graph
        for (int i = 0; i < n; i++) {
            equation = equations.get(i);
            
            // insert edge
            adjList = graph.get(equation.get(0));
            adjList.put(equation.get(1), values[i]);
            
            // insert reverse edge
            adjList = graph.get(equation.get(1));
            adjList.put(equation.get(0), 1/values[i]);
        }
        
        // run through queries
        HashSet<String> visited = new HashSet<String>();
        double[] sol = new double[queries.size()];
        List<String> query = new ArrayList<String>();
        double value;
        for(int i = 0; i < queries.size(); i++) {
            query = queries.get(i);
            if (!vertices.contains(query.get(0)) || !vertices.contains(query.get(0))) {
                sol[i] = -1.0;
            } else {
                value = dfs(graph, visited, query.get(0), query.get(1));
                sol[i] = value;
            }
        }
        
        return sol;
    }
    
    public double dfs(Map<String, Map<String, Double>> graph, 
                      HashSet<String> visited, 
                      String start,
                      String end) {
        // edge case
        if (!graph.containsKey(start)) {
            return -1.0;
        }
        
        // get the adj list
        Map<String, Double> adjList = graph.get(start);
        if (adjList.containsKey(end)) {
            return adjList.get(end);
        }
        
        // check neighbouring vertices
        visited.add(start);
        double sol = -1.0;
        for (String neigh: adjList.keySet()) {
            if (visited.contains(neigh)) {
                continue;
            }
            
            sol = dfs(graph, visited, neigh, end);
            if (sol != -1) {
                sol *= adjList.get(neigh);
                
                // save solution
                adjList.put(end, sol);
                break;
            }
        }
        
        visited.remove(start);
        return sol;
    }
}
